File: matlab/ch06/ex610_1.m
% sscstr.m: steady-state solution for a CSTR 
clear all; 
Caf = 10; D = 2.335; Fi = 10; Tf = 25; Tj = 25; dH = 5960; A1 = 3.49308e7;  
E = 11843; rCp = 500; Ui = 70; R = 1.987; 
Sa = (pi/4)*D^2; hs = Fi^2 / (10*Sa); Sh = Sa + pi*D*hs;  
a1 = Fi/Sa/hs; b1 = dH/rCp; c1 = Ui*Sh/(rCp*Sa*hs); 
fs = @(X) [a1*(Caf-X(1))-A1*exp(-E/R/(X(2)+273.15))*X(1); % X(1) = Cas, X(2) = Ts    
a1*(Tf - X(2)) + b1*A1*exp(-E/R/(X(2)+273.15))*X(1) - c1*(X(2) - Tj)]; 
% Solve nonlinear equations for three different initial guesses. 
x01 = [8 20]; x1 = fsolve(fs,x01); x02 = [5 70]; x2 = fsolve(fs,x02); x03 = [2 120]; x3 = fsolve(fs,x03); 
fprintf('Initial guess: Cas = %g, Ts = %g',x01(1), x01(2)); 
fprintf('  Steady-state values: Cas = %g, Ts = %g\n',x1(1), x1(2)); 
fprintf('Initial guess: Cas = %g, Ts = %g',x02(1), x02(2)); 
fprintf('  Steady-state values: Cas = %g, Ts = %g\n',x2(1), x2(2)); 
fprintf('Initial guess: Cas = %g, Ts = %g',x03(1), x03(2)); 
fprintf('  Steady-state values: Cas = %g, Ts = %g\n',x3(1), x3(2)); 
% Generate heat and temperature profiles. 
Ts = [20:0.1:120]; k1s = A1*exp(-E/R./(Ts+273.15)); Cas = a1*Caf./(k1s+a1); 
Qr = Ui*Sh*(Ts-Tj) + Fi*rCp*(Ts-Tf); Qg = dH*Sa*hs*Cas.*k1s; 
Tjs = Ts + (rCp*Fi*(Ts-Tf) - dH*Sa*hs*Cas.*k1s)/(Ui*Sh); 
subplot(1,2,1) % Heat profile 
plot(Ts,Qr,':',Ts,Qg), xlabel('Reactor temp.(deg.C)'), ylabel('Q(kcal/h)') 
legend('Qr','Qg','location','best'), grid 
subplot(1,2,2) % Temperature profile 
plot(Tjs,Ts), axis([0 60 25 110]), grid, xlabel('T_j(deg.C)'), ylabel('T(deg.C)') 

File: python/ch06/ex610_1.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import fsolve

# 1. 데이터 및 파라미터 설정
Caf = 10; D = 2.335; Fi = 10; Tf = 25; Tj = 25; dH = 5960; A1 = 3.49308e7
E = 11843; rCp = 500; Ui = 70; R = 1.987

# 2. 고정된 액위(hs) 및 관련 계수 계산
Sa = (np.pi / 4) * D**2
hs = Fi**2 / (10 * Sa)
Sh = Sa + np.pi * D * hs
a1 = Fi / Sa / hs
b1 = dH / rCp
c1 = Ui * Sh / (rCp * Sa * hs)

# 3. 정상 상태 방정식 정의
def sscstr(X):
    Cas, Ts = X  # X[0] = 농도, X[1] = 온도(deg.C)
    k1 = A1 * np.exp(-E / (R * (Ts + 273.15)))
    
    # f1: 농도 수지, f2: 에너지 수지
    f1 = a1 * (Caf - Cas) - k1 * Cas
    f2 = a1 * (Tf - Ts) + b1 * k1 * Cas - c1 * (Ts - Tj)
    return [f1, f2]

# 4. 세 가지 서로 다른 초기값으로 해 구하기
guesses = [[8, 20], [5, 70], [2, 120]]
solutions = []

print("CSTR Steady-State Analysis:")
for x0 in guesses:
    sol = fsolve(sscstr, x0)
    solutions.append(sol)
    print(f"Initial guess: Cas = {x0[0]:g}, Ts = {x0[1]:g}", end="")
    print(f"  -> Steady-state: Cas = {sol[0]:.4f}, Ts = {sol[1]:.2f}")

# 5. 열 및 온도 프로필 생성 (그래프용 데이터)[cite: 5]
Ts_range = np.arange(20, 120.1, 0.1)
k1s = A1 * np.exp(-E / (R * (Ts_range + 273.15)))
Cas_range = a1 * Caf / (k1s + a1)

# 제거되는 열(Qr)과 발생되는 열(Qg)[cite: 5]
Qr = Ui * Sh * (Ts_range - Tj) + Fi * rCp * (Ts_range - Tf)
Qg = dH * Sa * hs * Cas_range * k1s

# 자켓 온도(Tjs) 프로필 계산[cite: 5]
Tjs = Ts_range + (rCp * Fi * (Ts_range - Tf) - dH * Sa * hs * Cas_range * k1s) / (Ui * Sh)

# 6. 결과 시각화[cite: 5]
plt.figure(figsize=(12, 5))

# 왼쪽 그래프: 열 프로필 (S-곡선 확인)[cite: 5]
plt.subplot(1, 2, 1)
plt.plot(Ts_range, Qr, ':', label='Qr (Heat Removal)')
plt.plot(Ts_range, Qg, '-', label='Qg (Heat Generation)')
plt.xlabel('Reactor temp.(deg.C)')
plt.ylabel('Q(kcal/h)')
plt.title('Heat Generation vs removal')
plt.legend()
plt.grid(True)

# 오른쪽 그래프: Tj에 따른 Reactor Temperature[cite: 5]
plt.subplot(1, 2, 2)
plt.plot(Tjs, Ts_range)
plt.axis([0, 60, 25, 110])
plt.xlabel('$T_j$(deg.C)')
plt.ylabel('T(deg.C)')
plt.title('Hysteresis Loop (T vs Tj)')
plt.grid(True)

plt.tight_layout()
plt.show()